# Lab 1: Python Environments, JupyterLab, VS Code and Git

Name: MAINZA MUYA  
Student number:202406802   
Course: ICD341 Advanced Statistical Computing

## Part C: JupyterLab cells

### Cell 1: Python information

In [1]:
import sys
print(sys.version)
print(sys.executable)

3.12.5 (tags/v3.12.5:ff3bc82, Aug  6 2024, 20:45:27) [MSC v.1940 64 bit (AMD64)]
C:\Users\Hp\AppData\Local\Programs\Python\Python312\python.exe


### Cell 2: Import packages

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

print("Scientific Python environment working!")

Scientific Python environment working!


## Task 3: `sys.version` vs `sys.executable`

- `sys.version` is a **string describing the Python interpreter that is running**: its version number, build date and compiler.
- `sys.executable` is the **file path of the interpreter binary** that is running your code (for example `.../envs/icd341/python.exe`). It tells you *which* Python, and therefore which environment, is in use.

Together they let you confirm you are using Python 3.12 from the `icd341` environment.

## Task 4: VS Code vs JupyterLab

_Paste your two `sys.executable` outputs here (one from JupyterLab, one from VS Code) and confirm they match._

- JupyterLab: `...`
- VS Code: `...`

Both point to the `icd341` environment, so they share the same interpreter and packages.

## Part A: Questions
1. **What is a Python environment?** An isolated installation of Python plus a set of packages, kept separate from other installations on the machine.
2. **Why a separate environment for a course?** It avoids version conflicts with other projects and makes the setup reproducible and easy to rebuild or share.
3. **What does `conda activate icd341` do?** It switches the shell to use that environment's Python and packages, by putting its directories first on the PATH.
4. **Why `python=3.12`?** It pins the Python version so the environment behaves consistently, rather than accepting whatever default conda chooses.

## Part B: Questions
1. **`import numpy as np`** loads the NumPy package and lets you refer to it by the short name `np`.
2. **Why aliases?** They save typing and follow a community convention (`np`, `pd`, `plt`), which makes code easier to read.
3. **Importing a package that is not installed** raises `ModuleNotFoundError: No module named '...'`.

## Lab 1 short-answer questions
1. **Python environment:** an isolated set of a Python interpreter and installed packages.
2. **Two advantages of Conda environments:** (i) isolation, so different projects can use different package versions without clashing; (ii) reproducibility, because environments can be exported and recreated.
3. **JupyterLab:** a browser-based interface for notebooks, code, text, terminals and files, used for interactive and exploratory work.
4. **VS Code:** a free, extensible code editor from Microsoft with debugging, Git integration, and Python and notebook support.
5. **Git:** a distributed version-control system that tracks changes to files over time.
6. **GitHub:** a web platform that hosts Git repositories for backup, collaboration and sharing.
7. **`git add` vs `git commit`:** `git add` stages changes (chooses what goes into the next snapshot); `git commit` records the staged changes as a permanent snapshot with a message.
8. **Purpose of `.gitignore`:** it tells Git which files and folders not to track, such as caches, secrets and generated files.
9. **Why exclude `.ipynb_checkpoints`:** these are auto-generated, machine-specific backups that change constantly, add clutter and cause merge conflicts, and contain no original work.
10. **Why reproducible environments matter:** results depend on Python and package versions. A reproducible environment lets you or others rerun the analysis and obtain the same results, which is essential for scientific validity.

**Environment vs repository:** an environment holds the *software* that runs your code; a Git repository holds your *files and their history*.